flink-core/src/main/java/org/apache/flink/util/JavaGcCleanerWrapper.java

```diff
@@ -43,326 +47,338 @@
+		private final String cleanerName;
+		private final ReflectionUtils reflectionUtils;
+		private final String cleanerClassName;
+		private final Supplier<Optional<Object>> cleanerSupplier;
+		private final String cleanableCreationMethodName;
+		private final String cleanableClassName;
+		private final String cleanMethodName;
+
+		private CleanerFactoryProvider(
+				String cleanerName,
+				ReflectionUtils reflectionUtils,
+				String cleanerClassName,
+				Supplier<Optional<Object>> cleanerSupplier,
+				String cleanableCreationMethodName, // Cleaner is a factory for Cleanable
+				String cleanableClassName,
+				String cleanMethodName) {
+			this.cleanerName = cleanerName;
+			this.reflectionUtils = reflectionUtils;
+			this.cleanerClassName = cleanerClassName;
+			this.cleanerSupplier = cleanerSupplier;
+			this.cleanableCreationMethodName = cleanableCreationMethodName;
+			this.cleanableClassName = cleanableClassName;
+			this.cleanMethodName = cleanMethodName;
 		}
 
-		private static Method getJavaLangRefAccessTryHandlePendingReference(Class<?> javaLangRefAccessClass) {
-			try {
-				return javaLangRefAccessClass.getMethod("tryHandlePendingReference");
-			} catch (NoSuchMethodException e) {
-				throw new FlinkRuntimeException("Failed to find Java legacy JavaLangRefAccess#tryHandlePendingReference method", e);
-			}
+		private CleanerFactory createCleanerFactory() {
+			Class<?> cleanerClass = reflectionUtils.findClass(cleanerClassName);
+			Method cleanableCreationMethod = reflectionUtils.findMethod(
+				cleanerClass,
+				cleanableCreationMethodName,
+				Object.class,
+				Runnable.class);
+			Class<?> cleanableClass = reflectionUtils.findClass(cleanableClassName);
+			Method cleanMethod = reflectionUtils.findMethod(cleanableClass, cleanMethodName);
+			return new CleanerFactory(
+				cleanerName,
+				cleanerSupplier.get().orElse(null), // static method of Cleaner class will be called to create Cleanable
+				cleanableCreationMethod,
+				cleanMethod);
 		}
 
 		@Override
 		public String toString() {
-			return "Legacy cleaner provider before Java 9 using " + LEGACY_CLEANER_CLASS_NAME;
+			return cleanerName + " factory provider using " + cleanerClassName;
 		}
 	}
 
-	private static final class LegacyCleanerFactory implements CleanerFactory {
-		private final Method cleanerCreateMethod;
-		private final Method cleanerCleanMethod;
-		private final Method getJavaLangRefAccessMethod;
-		private final Method tryHandlePendingReferenceMethod;
-
-		private LegacyCleanerFactory(
-				Method cleanerCreateMethod,
-				Method cleanerCleanMethod,
-				Method getJavaLangRefAccessMethod,
-				Method tryHandlePendingReferenceMethod) {
-			this.cleanerCreateMethod = cleanerCreateMethod;
-			this.cleanerCleanMethod = cleanerCleanMethod;
-			this.getJavaLangRefAccessMethod = getJavaLangRefAccessMethod;
-			this.tryHandlePendingReferenceMethod = tryHandlePendingReferenceMethod;
+	private static class CleanerFactory {
```

Comment:
You don't have to change it back, but I'm a bit unsure whether unifying the `CleanerFactories` for Java 8 and > 8 really helped to improve maintainability. It is true that it reduced code duplication. However, it made it also much harder to see which methods are actually being called. In this case I would argue that simplifying the level of abstraction would have been better for maintainability (at the cost of a slight code duplication). Just ask yourself whether this version is easier to maintain when you look back at it in a month or so.

Reply:
Similar thoughts came into my mind. I think both approaches have benefits and drawbacks. Having less code makes it fast to understand what generally happens in both cases. True, arguably, the difference may be more obvious if you just compare two pieces of code. For this reason, I left comments for Java 8 case explaining the difference. It is also a good idea to add the normal Java code for both cases, I can do this.